src/Microsoft.AspNetCore.Proxy/ProxyExtensions.cs

```diff
@@ -0,0 +1,64 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.Proxy;
+using Microsoft.Extensions.DependencyInjection;
+
+namespace Microsoft.AspNetCore.Builder
+{
+    public static class ProxyExtensions
+    {
+        /// <summary>
+        /// Sends request to the specified server
+        /// </summary>
+        /// <param name="app"></param>
+        /// <param name="destinationUri">Destination Uri</param>
+        public static void RunProxy(this IApplicationBuilder app, Uri destinationUri)
+        {
+            if (destinationUri == null)
+            {
+                throw new ArgumentNullException(nameof(destinationUri));
+            }
+            app.Run(ctx => ctx.ProxyRequest(destinationUri));
+        }
+
+        /// <summary>
+        /// Sends request to the specified server
+        /// </summary>
+        /// <param name="context"></param>
+        /// <param name="destinationUri">Destination Uri</param>
+        public static async Task ProxyRequest(this HttpContext context, Uri destinationUri)
+        {
+            if (destinationUri == null)
+            {
+                throw new ArgumentNullException(nameof(destinationUri));
+            }
+
+            if (context.WebSockets.IsWebSocketRequest)
+            {
+                await context.AcceptProxyWebSocketRequest(destinationUri.ToWebSocketScheme());
+            }
+            else
+            {
+                var proxyService = context.RequestServices.GetRequiredService<ProxyService>();
+
+                using (var requestMessage = context.CreateProxyHttpRequest(destinationUri))
+                {
+                    var prepareRequestHandler = proxyService.Options.PrepareRequest;
```

Comment:
This is a change introduced for #35 and #4 
Basically in my opinion for most of scenarios (including retries) DelegatingHandler should be used.
I have introduced this PrepareRequest handler in the last commit to allow to altering this request while having access to the original HttpRequest. I am not sure about this one. If you approve it I will add tests and examples.

If you reject it I am also OK with your decision, as for my scenarios DelegatingHandler is more than enough. And in extreme cases AsyncLocal can be used to pass HttpContext to the handler.

So what's your verdict here? I am OK with both.

Regarding #4 I think we should built this in. But I will do this in a separate pull request it deserves a separate discussion. It's quite complex as there is an emerging HTTP specification called Forwarded, and competitive a competitive one defacto standard X-Forwarded-Host... You can learn more here #Azure/service-fabric-issues/issues/113 .